JBPM-6031: Stunner - Toolbox improvements.

romartin opened the pull request:
Hey @manstis 

Well this is my last commit before going on PTO... and well.. it's quite big, although it's completely related to toolbox stuff (and the canvas control for it too).

This commit:
- Removes the actual toolbox implementation and replaces it by a new one. There were no tests at all before, so I've added most of them too for the new stuff. Notice all toolbox stuff is placed in the `lienzo-extensions` module, as it's pure lienzo stuff that does not depend on uf/errai, etc.
- Removed all unnecessary toolbox API in Stunner... just a few interfaces now, big part is pure view (lienzo).
- Refactored `ToolboxCanvasControl`and replaced `ToolboxCommand`stuff  by `ToolboxAction`s - the new implementation are much shorter, better and easy to read & understand.
- Removed too all the classes related to the `LienzoPalette` and its views, there were used by the morphing toolbox but actually the toolbox itself provides that drop-down stuff, so as nobody uses the LienzoPalette stuff now, removed it.
- As found some bugs about  selection control, I've done some fixes too.
- Refactored DMN toolbox stuff

PS: @manstis  I'm still working on a few missing tests for the new lienzo toolbox classes, but would be nice, as @hasys  is on PTO this week, if you can give it a try and provide some feedback pleeease, this way tomorrow I can add the missing tests and fix whatever feedback you can provide, and we're done for Friday... :)

Thanks!!!

manstis commented:
<cough> 208, <cough>, 208 files ;-)

manstis commented:
Jenkins please execute full downstream build.

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/canvas/controls/toolbox/DMNToolboxControl.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Inject
+    public DMNToolboxControl(final @DMNFlowActionsToolbox ActionsToolboxFactory flowActionsToolboxFactory,
+                             final @CommonActionsToolbox ActionsToolboxFactory commonActionsToolboxFactory) {
```
I think this is where the missing "Edit Decision" Toolbox item goes missing. ```@CommonActionsToolbox``` only contains "delete". So this probably needs to use a specific implementation for DMN or add another ```ActionsToolboxFactory``` for the missing item but IDK how composition with multiple factories works.

romartin replied:
oh yes! thanks for pointing this, just adding it, and sorry!

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/canvas/controls/toolbox/DMNToolboxControlTest.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Test
+    public void testRegisterTheRightFactories() {
```
This will need changing when the missing "Edit Decision" Toolbox Item is added back.

romartin replied:
not sure if this test makes sense... as it only should test the right CDI injections.... the ToolboxControlImpl, which this one delegates to, is already tested too.

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/toolbox/AbstractToolboxControl.java:
```diff
@@ -0,0 +1,114 @@
+        return toolboxControl.getToolboxes(element);
+    }
+
+    void onCanvasElementSelectedEvent(final @Observes CanvasElementSelectedEvent event) {
```
I ran into a problem myself with classes observing things yesterday. If I want to override these methods in a sub-class the sub-class needs to also observe the events and then I found CDI (or at least Errai) was running both observers. I had to change ```SessionPreviewImpl``` to delegate handling of the event to a ```protected``` method and then I'd override the new method and not the "observing" method. See [here](https://github.com/kiegroup/kie-wb-common/pull/1068/files#diff-925cc409d06c3ed587ef0f73a28153ed) for an example.

manstis replied:
IDK if you want to consider the same here to avoid future potential (difficult to diagnose) problems?

romartin replied:
sure ,lemme give it a look! :)

romartin replied:
mmm good finding! was not aware of this, thanks for the feedback, one new thing learned today :) refactoring this

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/components/toolbox/actions/CreateConnectorAction.java:
```diff
@@ -0,0 +1,316 @@
+        canvasHighlight = canvasHighlightBuilder.apply(canvasHandler);
+        graphBoundsIndexer.setRootUUID(canvasHandler.getDiagram().getMetadata().getCanvasRootUUID());
+        graphBoundsIndexer.build(canvasHandler.getDiagram().getGraph());
+        edgeBuilderControl.enable(canvasHandler);
```
Out of interest why do you have to enable/disable the ```EdgeBuilderControl``` and set/clear its ```CommandManager``` on ```start(..)``` and ```complete(...)``` methods?

romartin replied:
good point ,not sure tbh, probably can be improved, looking :)

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/components/toolbox/actions/CreateNodeAction.java:
```diff
@@ -0,0 +1,208 @@
+        return this;
+    }
+
+    private void start(final AbstractCanvasHandler canvasHandler) {
```
Same comment about start/complete.

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-lienzo-extensions/src/main/java/org/kie/workbench/common/stunner/lienzo/toolbox/GroupItem.java:
```diff
@@ -0,0 +1,119 @@
+        return this;
+    }
+
+    public GroupItem show(final Runnable before,
```
I noticed your move to ```Runnable``` in this PR. Why not ```Command``` like we (me?) normally uses. Client-side code has no concept of a ```Runnable``` in the same sense as regular Java code.

romartin replied:
oh yes! can refactor it, np! There exist a reason for this, I did all the toolbox development in `my-lienzo-tests`, for being faster, and that's working still in java7/gwt.27, no errai/uf... so I used `Runnable`, but now that's integrated into Stunner, can replace it by `Command` yeah, makes sense.

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-lienzo-extensions/src/main/java/org/kie/workbench/common/stunner/lienzo/toolbox/grid/AutoGrid.java:
```diff
@@ -0,0 +1,197 @@
+import com.ait.lienzo.client.core.types.Point2D;
+import com.ait.lienzo.shared.core.types.Direction;
+
+public class AutoGrid extends AbstractLayoutGrid<AutoGrid> implements SizeConstrainedGrid<Point2D> {
```
IDK.. wasn't some of this shown in your Lienzo Kitchen Sink demo of the new Toolbox? Aren't these classes in Lienzo?

romartin replied:
that's the point, these classes are not yet in lienzo, because it's java8... I created the classes in `lienzo-extensions` stunner's module, so once lienzo is moved to `3.0`. this can be moved to... EDYT?

romartin replied:
so I created the classes initially in `my-lienzo-tests`, and finally moved to `lienzo-extensions` (stunner)

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-lienzo-extensions/src/main/java/org/kie/workbench/common/stunner/lienzo/toolbox/items/decorator/DecoratorsFactory.java:
```diff
@@ -0,0 +1,44 @@
+
+public class DecoratorsFactory {
+
+    public static DecoratorsFactory INSTANCE = new DecoratorsFactory();
```
If you have a static (singleton) instance you might want a ```private``` constructor?

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-lienzo-extensions/src/main/java/org/kie/workbench/common/stunner/lienzo/toolbox/items/impl/AbstractFocusableGroupItem.java:
```diff
@@ -0,0 +1,175 @@
+        useFocusGroupExecutor(new FocusGroupExecutor());
+    }
+
+    T focus() {
```
I notice elsewhere, if the timer is already running you cancel it first. No need here?

romartin replied:
mmmmmm

romartin replied:
yep I think it was a bug, adding the check and testing it. Thx!

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-lienzo-extensions/src/main/java/org/kie/workbench/common/stunner/lienzo/toolbox/items/impl/AbstractFocusableGroupItem.java:
```diff
@@ -0,0 +1,175 @@
+        return cast();
+    }
+
+    T unFocus() {
```
I notice elsewhere, if the timer is already running you cancel it first. No need here?

romartin replied:
same as above too :)

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-lienzo-extensions/src/main/java/org/kie/workbench/common/stunner/lienzo/toolbox/items/impl/ButtonGridItemImpl.java:
```diff
@@ -0,0 +1,389 @@
+                registerItemUnFocusHandler(primitiveItem,
+                                           itemUnFocusCallback);
+            } catch (final ClassCastException e) {
+                throw new UnsupportedOperationException("The button only supports subtypes " +
```
Then why not make method signature ```AbstractDecoratedItem... items```?

romartin replied:
well yes... this probably indicates some lack about right use of generics for the "decorators" stuff. 
Actually this is because this signature is given by the interface `ButtonGridItem`, which implements `ItemGrid` but using the `DecoratedItem` type. 
The idea is that the interfaces, like `DecoratedItem`, `ButtonItem`, `ButtonGridItem`, etc, just provide the methods to handle the item's behavior, like `show`, `decorate`, etc, but the interfaces are not aware if that items are primitives and how those can be added into the layer. Then the `AbstracXXX` classes, like `AbstractPrimitiiveItem`, `AbstractDecoratorItem`, etc are the ones that provide the `asPrimitive()` method, and all `AbstractXXX` implementations can handle items that are extended from that abstractions or not, this way that classes can be extended, the item behviours are handled by this abstractions, but you can provide a different way to be added into the canvas/layer, or even not being a primitive... 
But yeah, your comment makes sense, now looking at this it sounds a bit strange, although tbh don't think can refactor this on time :/ if you think so can add a `TODO` here plz, so we can address this later on....

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-lienzo-extensions/src/main/java/org/kie/workbench/common/stunner/lienzo/toolbox/items/impl/ButtonGridItemImpl.java:
```diff
@@ -0,0 +1,389 @@
+    }
+
+    private boolean isDropDown() {
+        return arrow.getRotationDegrees() == 0;
```
It might have been nicer to preserve ```Direction``` from the ```at(..)``` method to keep track of the _dropdown_ orientation.. but a minor comment.

romartin replied:
oh sure well seen, absolutely better and nicer, refactoring it, it's just a bit

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-lienzo-extensions/src/main/java/org/kie/workbench/common/stunner/lienzo/toolbox/items/impl/ButtonItemImpl.java:
```diff
@@ -0,0 +1,144 @@
+        return this;
+    }
+
+    public ButtonItemImpl onDragStart(final NodeDragStartHandler handler) {
```
So all buttons track it's parent shape drag separately and move themselves? I mean they are not attached to the same ```Group``` as the parent shape and hence drag automagically? Not a problem just curious!

romartin replied:
Well yes, all buttons are attached to the `Group` as the parent shape, and if you move the "toolbox" or the "item grid", so if you move the parent `Group`, the buttons drag automatically. 
But this is to provide both `click` and `drag` event support for the toolbox' buttons. So for each button, you can just click on it, or drag it, this way the target location is not "automatically" discovered, it's the one where user drops.
Notice the lienzo toolbox' buttons support both click and drag, as you can see in this class (or any  `ActionItem` implementation ), but the drag behavior is not still implemented/supported in Stunner side, for now you can just click on the button and the new shape gets created on some next "available" location. I expect adding drag support soon.. but stunner still does not use it.
Makes sense all this? Thanks for the comments/feedback!

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-lienzo-extensions/src/main/java/org/kie/workbench/common/stunner/lienzo/toolbox/items/impl/ButtonsFactory.java:
```diff
@@ -0,0 +1,77 @@
+
+public class ButtonsFactory {
+
+    public static ButtonsFactory INSTANCE = new ButtonsFactory();
```
You might want a ```private``` constructor if you want to rely on use of a single instance.

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-lienzo-extensions/src/main/java/org/kie/workbench/common/stunner/lienzo/toolbox/items/impl/ToolboxFactory.java:
```diff
@@ -0,0 +1,42 @@
+
+public class ToolboxFactory {
+
+    public static ToolboxFactory INSTANCE = new ToolboxFactory();
```
You might want a ```private``` constructor.

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-lienzo-extensions/src/main/java/org/kie/workbench/common/stunner/lienzo/toolbox/items/tooltip/TooltipFactory.java:
```diff
@@ -0,0 +1,35 @@
+
+public class TooltipFactory {
+
+    public static TooltipFactory INSTANCE = new TooltipFactory();
```
Again, consider a ```private``` constructor.

romartin commented:
hey @manstis , thanks for the review!! So just looking at the missing button for DMN (sorry about t that, promised I tested it, but probably didn't noticed about this button :( ...) and fixing the comments above. Will let you know soon. 
Thanks!

romartin commented:
@manstis okk great, just pushed a new commit - second iteration hehe. Fixed some of the comments above (added some answers for others) and added the lost toolbox button for edit DMN decisions! Sorry, completely forgot it! 
Would like to add the missing tests for some of the lienzo tooblox classes yet, just working on this again, but rather than that, I think almost done.. :)

manstis commented:
@romartin Thanks for the changes (when I manage to find them).

I've found there is really no need to squash and (force) push updates; just add another commit and they can be squashed by Guithub when merging by the Gate Keeper. The commit comment combined all of the individual commit messages so can look a little strange but it'd make reviewing your changes to this PR soooooo much easier :-)

romartin commented:
@manstis  okkk yeah sure! easier for everyone hehe, will do!

manstis commented:
@romartin I know I can always call on [you](https://www.youtube.com/watch?v=L_fCqg92qks)!

manstis commented:
(fat fingers)

manstis commented:
Let Jenkins deal with it!

romartin commented:
yeah my friend J

romartin commented:
failed due to "No space left on device" :O

manstis commented:
@romartin I now get this when running,,,,
![screenshot from 2017-08-10 23-33-03](https://user-images.githubusercontent.com/497544/29195041-612edc76-7e24-11e7-97cb-d8fc03752a0c.png)

romartin commented:
@manstis  oh I did test this, it's working for me, anyway lemme try it again in another laptop

romartin commented:
@manstis this happens opening an existing diagram or just creating a new one?

manstis commented:
Creating new. I just fetched the PR and ran webapp then tried creating a
new diagram. Late now.. won't try again until tomorrow.

romartin commented:
@manstis  sure no worries, just testing this in a "clean" environment too, will let you know! thanks!!!

romartin commented:
it works for me on another laptop, anyway trying to figure out from the screenshot if can be some issue

romartin commented:
Hey @manstis 
Just added the missing tests in a new commit! Also I did different runs for the DMN showcase and ir works for me..,  can you please confirm tomorrow??
Thanks!!

manstis commented:
@romartin All works good.. and looks great!

romartin commented:
@manstis  not sure if the build will be completed this week eh :/ it should be compiling the linux kernel too :P

romartin commented:
Jenkins please retest this

romartin commented:
oh looks like a findbugs issue, build failed, just looking

romartin commented:
@manstis fixed the findbugs issue, forgot to run the full build locally... sorry! 
CC @hasys nice! it was a potential bug ;)

romartin commented:
Jenkins please retest this

romartin commented:
@manstis realized that a couple of icons were not being rendered on the canvas in FF. Just fixed too. Thanks!

romartin commented:
Jenkins please retest this
--
